frontend/src/app/main/ui/shapes/fills.cljs

```diff
@@ -20,45 +20,54 @@
-       (for [[index value] (-> (d/enumerate (:fills shape [])) reverse)]
-         (cond (some? (:fill-color-gradient value))
-           (case (:type (:fill-color-gradient value))
-             :linear [:> grad/linear-gradient #js {:id (str (name :fill-color-gradient) "_" render-id "_" index)
-                                                   :gradient (:fill-color-gradient value)
-                                                   :shape shape}]
-             :radial [:> grad/radial-gradient #js {:id (str (name :fill-color-gradient) "_" render-id "_" index)
-                                                   :gradient (:fill-color-gradient value)
-                                                   :shape shape}])))
+    (when (or (some? (:fill-image shape))
+              (#{:image :text} (:type shape))
+              (> (count (:fills shape)) 1)
+              (some :fill-color-gradient (:fills shape)))
 
-       [:> :pattern pattern-attrs
-        [:g
-         (for [[index value] (-> (d/enumerate (:fills shape [])) reverse)]
-           [:> :rect (-> (attrs/extract-fill-attrs value index)
-                         (obj/set! "width" width)
-                         (obj/set! "height" height))])
+      (let [{:keys [x y width height]} (:selrect shape)
+            {:keys [metadata]} shape
+            
+            has-image (or metadata (:fill-image shape))
+            uri (if metadata
+                  (cfg/resolve-file-media metadata)
+                  (cfg/resolve-file-media (:fill-image shape)))
+            embed (embed/use-data-uris [uri])
+            transform (gsh/transform-matrix shape)
+            pattern-attrs (cond-> #js {:patternUnits "userSpaceOnUse"
+                                       :x x
+                                       :y y
+                                       :height height
+                                       :width width
+                                       :data-loading (str (not (contains? embed uri)))}
+                            (= :path (:type shape))
+                            (obj/set! "patternTransform" transform))]
 
-         (when has-image
-           [:image {:xlinkHref (get embed uri uri)
-                    :width width
-                    :height height}])]]]))
+        [:*
+         (for [[_shape-index shape] (d/enumerate (or (:position-data shape) [shape]))]
+           (for [[fill-index value] (-> (d/enumerate (:fills shape [])) reverse)]
+             (cond (some? (:fill-color-gradient value))
+                   (case (d/name (:type (:fill-color-gradient value)))
+                     "linear" [:> grad/linear-gradient #js {:id (str "fill-color-gradient_" render-id "_" fill-index)
+                                                            :gradient (:fill-color-gradient value)
+                                                            :shape shape}]
+                     "radial" [:> grad/radial-gradient #js {:id (str "fill-color-gradient_" render-id "_" fill-index)
```

Comment:
I'm pretty sure that in this case, where the data is known statically, the `#js` prefix is not necessary, but check it before just trusting me xD

Reply:
I think this was supposed to be `:>` so the props can be reused and not repeat code. I'll change so both component use the same props